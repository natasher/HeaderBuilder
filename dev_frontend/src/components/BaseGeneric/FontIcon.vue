<template>
  <i
    :class="classObject"
    aria-hidden="true">
  </i>
</template>

<script>
export default {

  name: 'FontIcon',

  props: {
    /**
     * Icon name from Font Awesome cataloge.
     */
    icon: {
      type    : String,
      required: true,
    },
  },

  computed: {
    classObject: function () {
      return `icon-${ this.icon }`
    },
  },
}
</script>
