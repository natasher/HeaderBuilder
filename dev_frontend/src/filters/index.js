/**
 * Capitalize first letter of str parameter.
 * e.g.
 *
 * "foo" => "Foo"
 * @param {string} str
 */
export const capitalize = ( value ) => {
  if ( !value ) return ''

  value = value.toString()
  return value.charAt( 0 ).toUpperCase() + value.slice( 1 )
}

/**
 * Convert string with dash separated words (e.g. value of data-type param)
 * to space separated string.
 * e.g.
 *
 * "menu_icon" => "menu icon"
 * @param {string} str
 */
export const convertDashToSpace = ( str ) => {
  if ( !str ) return ''

  return str.split( '_' ).join(' ')
}

/**
 * Convert string with space separated words to dash separated string.
 * e.g.
 *
 * "Action Bar" => "Action_Bar"
 * @param {string} str
 */
export const convertSpaceToDash = ( str ) => {
  if ( !str ) return ''

  return str.split(' ').join( '_' )
}

/**
 * Convert string with space to one Camel Cased string.
 * e.g.
 *
 * "background color" => "backgroundColor"
 * @param {string} str
 */
export const toCamelCase = ( str ) => {
  if ( !str ) return ''

  let arr    = str.split(' ')
  let output = arr.reduce(( prev, curr, index ) => {
    // return ( index === 0 ) ? prev + curr : prev + capitalize( curr )
    return (index === 0)
      ? curr.charAt(0).toLowerCase() + curr.slice(1)
      : prev + capitalize(curr)
  }, '')

  return output
}

/**
 * Convert camelCased string to human-readable string.
 * e.g.
 *
 * "backgroundColor" => "Background Color"
 * @param {string} str
 */
export const unCamelCase = ( str ) => {
  if ( !str ) return ''

  return str
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .replace(/\b([A-Z]+)([A-Z])([a-z])/, '$1 $2$3')
    .replace(/^./, ( str ) => { return str.toUpperCase() })
}

/**
 * Convert camelCased string to upper-cased snake case.
 * e.g.
 *
 * "secondRow" => "SECOND_ROW"
 */
export const allCaps = ( str ) => {
  if ( !str ) return ''

  return convertSpaceToDash(unCamelCase( str )).toUpperCase()
}