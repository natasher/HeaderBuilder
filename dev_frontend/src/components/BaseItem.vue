<template>
  <div
    class        = "hb-item"
    :class       = "{'hb-item-placeholder': placeholder}"
    :data-type   = "item.name"
    :data-cellId = "cellId">

    <span class = "icons">

      <a class = "icon-edit">
        <i
          class          = "icon-pencil"
          @click.passive = "showForm">
        </i>
      </a>

      <a class = "icon-delete">
        <i
          class          = "icon-cancel"
          @click.passive = "destroyItem">
        </i>
      </a>

    </span>

    <div
      class  = "image"
      v-show = "!placeholder">
        <font-icon :icon="itemIcon" />
    </div>

    <span
      class  = "title"
      v-show = "!placeholder">
        {{ item.name | unCamelCase }}
    </span>

  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex'
import { unCamelCase } from '../filters/index'
import FontIcon from './BaseGeneric/FontIcon.vue'

export default {

  name: 'BaseItem',

  components: {
    FontIcon
  },

  filters: {
    unCamelCase
  },

  props: {
    /**
     * Pass `Item` object which instance will represent.
     * The `Item` must have at least two properties: name & icon.
     * The `item.form` is taken from store editor grid,
     * based on other passed props: `row`, `cellPos`, `cellId`.
     */
    item: {
      type    : Object,
      required: true,
    },
    /**
     * cellId - item `id` in grid-cell array (from left, starts at 0)
     * If non is given, the item is in the `ItemsTray`.
     */
    cellId: {
      type    : Number,
      required: false,
      default : void 0,
    },
    /**
     * In which row the `Item` lays.
     * If non is given, the item is in the `ItemsTray`.
     *    [ 'actionBar', 'fristRow', 'secondRow' ]
     */
    row: {
      type    : String,
      required: false,
      default : void 0,
    },
    /**
     * In which cell in row, the `Item` lays.
     * If non is given, the item is in the `ItemsTray`.
     *    [ 'all', 'left', 'center', 'right' ]
     */
    cellPos: {
      type    : String,
      required: false,
      default : void 0,
    },
    /**
     * Used during item dragging.
     * Show only grayed rounded square.
     */
    placeholder: {
      type   : Boolean,
      require: false,
      default: false,
    },
  },

  methods: {
    ...mapActions( 'modals', [
      'showModalItemForm',
    ]),

    ...mapActions( 'cell', [
      'removeItem',
    ]),

    ...mapActions( 'uuid', [
      'removeFromBlacklist',
    ]),

    showForm () {
      if ( this.getGridStatus === 'custom' ) {
        this.showModalItemForm({
          originalCoords: this.item.originalCoords,
          $modal        : this.$modal
        })
      }
    },

    destroyItem () {
      if ( this.getGridStatus === 'custom' ) {

        this.removeItem({
          cellId        : this.cellId,
          row           : this.row,
          cellPos       : this.cellPos,
          originalCoords: this.item.originalCoords,
          confirm       : true,
        })

        // dispatch( 'uuid/removeFromBlacklist', , { root: true })
        this.removeFromBlacklist({ itemUuid: this.item.uuid })
      }
    },
  },

  computed: {
    ...mapGetters( 'ui', [ 'getGridStatus' ]),

    itemIcon: function () {
      const iconsList = {
        logo    : 'wordpress',
        menu    : 'menu',
        menuIcon: 'menu',
        extras  : 'flickr',
        social  : 'share',
        text    : 'instapaper',
        image   : 'picture',
        icon    : 'feather',
        button  : 'progress-0',
      }

      return iconsList[ this.item.name ]
    },

  }

}
</script>