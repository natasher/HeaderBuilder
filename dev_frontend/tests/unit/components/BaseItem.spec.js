import { shallowMount, createLocalVue } from '@vue/test-utils'
import { unCamelCase } from '../../../src/filters/index'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import BaseItem      from '../../../src/components/BaseItem'
import FontIcon      from '../../../src/components/BaseGeneric/FontIcon.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      modals: {
        namespaced: true,
        actions: {
          showModalItemForm: jest.fn(() => Promise.resolve())
        }
      },
      cell: {
        namespaced: true,
        actions: {
          removeItem: jest.fn(() => Promise.resolve())
        }
      },
      ui: {
        namespaced: true,
        getters: {
          getGridStatus: jest.fn(() => { return 'custom' })
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    filters: { unCamelCase },
    store: createStore(),
    stubs: [
      'FontIcon',
    ],
    propsData: {
      item: {
        name: 'testName',
        icon: 'testIcon',
      },
    }
  }

  return shallowMount( BaseItem, merge( defaultMountingOptions, overrides ))
}

describe('BaseItem.vue', () => {

  describe('snapshots:', () => {

    test('with only required props', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('when item is dragged', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: true
        }
      })
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  describe('props:', () => {

    test('`item` should be an object with name and icon', async () => {
      const item = {
        name: 'someName',
        icon: 'crazyIcon'
      }
      const wrapper = createWrapper({
        propsData: {
          item
        }
      })
      await flushPromises()

      expect( wrapper.props().item ).toEqual( item )
    })

    test('`cellId` should exist and be empty by default', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.props().cellId ).toBe( void 0 )
    })

    test('`row` should exist and be empty by default', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.props().row ).toBe( void 0 )
    })

    test('`cellPos` should exist and be empty by default', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.props().cellPos ).toBe( void 0 )
    })

    test('`placeholder` should exist and be false by default', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.props().placeholder ).toBe( false )
    })

  })

  describe('methods:', () => {

    test('`showModalItemForm` should be called if grid status is `custom`', async () => {
      const showModalItemForm = jest.fn()
      const store = createStore({
        modules: {
          modals: {
            actions: {
              showModalItemForm,
            }
          },
          ui: {
            getters: {
              getGridStatus: () => { return 'custom' }
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      wrapper.find('a.icon-edit>i').trigger('click')
      await flushPromises()

      expect( showModalItemForm ).toBeCalled()
    })

    test('`showModalItemForm` should NOT be called if grid status is NOT `custom`', async () => {
      const showModalItemForm = jest.fn()
      const store = createStore({
        modules: {
          modals: {
            actions: {
              showModalItemForm,
            }
          },
          ui: {
            getters: {
              getGridStatus: () => { return 'other' }
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      wrapper.find('a.icon-edit>i').trigger('click')
      await flushPromises()

      expect( showModalItemForm ).not.toBeCalled()
    })

    test('`removeItem` should be called if grid status is `custom`', async () => {
      const removeItem = jest.fn()
      const store = createStore({
        modules: {
          cell: {
            actions: {
              removeItem,
            }
          },
          ui: {
            getters: {
              getGridStatus: () => { return 'custom' }
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      wrapper.find('a.icon-delete>i').trigger('click')
      await flushPromises()

      expect( removeItem ).toBeCalled()
    })

    test('`removeItem` should NOT be called if grid status is NOT `custom`', async () => {
      const removeItem = jest.fn()
      const store = createStore({
        modules: {
          cell: {
            actions: {
              removeItem,
            }
          },
          ui: {
            getters: {
              getGridStatus: () => { return 'other' }
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      wrapper.find('a.icon-delete>i').trigger('click')
      await flushPromises()

      expect( removeItem ).not.toBeCalled()
    })

  })

  describe('dynamic behaviour:', () => {

    test('should has class `hb-item-placeholder` if placeholder prop is true', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: true
        }
      })
      await flushPromises()

      expect( wrapper.classes() ).toContain( 'hb-item-placeholder' )
    })

    test('should NOT has class `hb-item-placeholder` if placeholder prop is true', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: false
        }
      })
      await flushPromises()

      expect( wrapper.classes() ).not.toContain( 'hb-item-placeholder' )
    })

    test('should has `data-type` attribute with the item name', async () => {
      const item = {
        name: 'someName',
        icon: 'crazyIcon'
      }
      const wrapper = createWrapper({
        propsData: {
          item
        }
      })
      await flushPromises()

      expect( wrapper.attributes()[ 'data-type' ] ).toBe( item.name )
    })

    test('should has `data-cellId` attribute with cellId', async () => {
      const cellId = 666
      const wrapper = createWrapper({
        propsData: {
          cellId,
        }
      })
      await flushPromises()

      expect( wrapper.attributes()[ 'data-cellid' ] ).toBe( `${ cellId }` )
    })

    test('should show item icon if `placeholder` is false', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: false
        }
      })
      await flushPromises()

      expect( wrapper.find('.image' ).isVisible() ).toBe( true )
    })

    test('should hide item icon if `placeholder` is true', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: true
        }
      })
      await flushPromises()

      expect( wrapper.find('.image' ).isVisible() ).toBe( false )
    })

    test('should show item title if `placeholder` is false', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: false
        }
      })
      await flushPromises()

      expect( wrapper.find('.title' ).isVisible() ).toBe( true )
    })

    test('should show item title if `placeholder` is false', async () => {
      const wrapper = createWrapper({
        propsData: {
          placeholder: true
        }
      })
      await flushPromises()

      expect( wrapper.find('.title' ).isVisible() ).toBe( false )
    })

    test('should render item name in readable format (unCamelCased)', async () => {
      const item = {
        name: 'someName',
        icon: 'crazyIcon'
      }
      const wrapper = createWrapper({
        propsData: {
          item
        }
      })
      await flushPromises()

      expect( wrapper.find('.title').text() ).toBe( 'Some Name' )
    })

    test('should pass `icon` as a prop to <font-icon />', async () => {
      const item = {
        name: 'someName',
        icon: 'crazyIcon'
      }
      const wrapper = createWrapper({
        propsData: {
          item
        }
      })
      await flushPromises()

      expect( wrapper.find( FontIcon ).props( 'icon' ) ).toEqual({ icon: item.icon })
    })

  })

})