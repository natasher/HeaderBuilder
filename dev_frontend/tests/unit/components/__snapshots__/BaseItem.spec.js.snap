// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseItem.vue snapshots: when item is dragged 1`] = `
<div
  class="hb-item hb-item-placeholder"
  data-type="testName"
>
  <span
    class="icons"
  >
    <a
      class="icon-edit"
    >
      <i
        class="icon-pencil"
      />
    </a>
     
    <a
      class="icon-delete"
    >
      <i
        class="icon-cancel"
      />
    </a>
  </span>
   
  <div
    class="image"
    style="display: none;"
  >
    <fonticon-stub
      icon="testIcon"
    />
  </div>
   
  <span
    class="title"
    style="display: none;"
  >
    
      Test Name
  
  </span>
</div>
`;

exports[`BaseItem.vue snapshots: with only required props 1`] = `
<div
  class="hb-item"
  data-type="testName"
>
  <span
    class="icons"
  >
    <a
      class="icon-edit"
    >
      <i
        class="icon-pencil"
      />
    </a>
     
    <a
      class="icon-delete"
    >
      <i
        class="icon-cancel"
      />
    </a>
  </span>
   
  <div
    class="image"
  >
    <fonticon-stub
      icon="testIcon"
    />
  </div>
   
  <span
    class="title"
  >
    
      Test Name
  
  </span>
</div>
`;
